core/src/consensus/tendermint/engine.rs

```diff
@@ -254,3 +292,261 @@ impl ConsensusEngine for Tendermint {
+    }
+
+    let mut reduced_rewards = 0;
+
+    // Penalty disloyal validators
+    let number_of_blocks_in_term = start_of_the_current_term - start_of_the_previous_term;
+    for (address, intermediate_reward) in rewards {
+        // FIXME: Consider banned accounts
+        let number_of_signatures = u64::try_from(*signed_blocks.get(&address).unwrap()).unwrap();
+        let final_block_rewards = final_rewards(intermediate_reward, number_of_signatures, number_of_blocks_in_term);
+        reduced_rewards += intermediate_reward - final_block_rewards;
+        pending_rewards.insert(address, final_block_rewards);
+    }
+
+    // Give additional rewards
+    give_additional_rewards(reduced_rewards, missed_signatures, |address, reward| {
+        pending_rewards.insert(*address, reward);
+        Ok(())
+    })?;
+
+    Ok(pending_rewards)
+}
+
+/// reward = floor(intermediate_rewards * (a * number_of_signatures / number_of_blocks_in_term + b) / 10)
+fn final_rewards(intermediate_reward: u64, number_of_signatures: u64, number_of_blocks_in_term: u64) -> u64 {
+    let (a, b) = if number_of_signatures * 3 >= number_of_blocks_in_term * 2 {
+        // number_of_signatures / number_of_blocks_in_term >= 2 / 3
+        // x * 3/10 + 7/10
+        (3, 7)
+    } else if number_of_signatures * 3 >= number_of_blocks_in_term {
+        // number_of_signatures / number_of_blocks_in_term >= 1 / 3
+        // x * 24/10 - 7/10
+        (24, -7)
+    } else {
+        // 1 / 3 > number_of_signatures / number_of_blocks_in_term
+        // x * 3/10 + 0
+        assert!(
+            number_of_blocks_in_term > 3 * number_of_signatures,
+            "number_of_signatures / number_of_blocks_in_term = {}",
+            (number_of_signatures as f64) / (number_of_blocks_in_term as f64)
+        );
+        (3, 0)
+    };
+    let numerator = i128::from(intermediate_reward)
+        * (a * i128::from(number_of_signatures) + b * i128::from(number_of_blocks_in_term));
+    assert!(numerator >= 0);
+    let denominator = 10 * i128::from(number_of_blocks_in_term);
+    // Rust's division rounds towards zero.
+    u64::try_from(numerator / denominator).unwrap()
+}
+
+fn give_additional_rewards<F: FnMut(&Address, u64) -> Result<(), Error>>(
+    mut reduced_rewards: u64,
+    missed_signatures: HashMap<Address, (usize, usize)>,
+    mut f: F,
+) -> Result<(), Error> {
+    let sorted_validators = missed_signatures
+        .into_iter()
+        .map(|(address, (proposed, missed))| (address, Ratio::new(missed, proposed)))
+        .fold(BTreeMap::<Ratio<usize>, Vec<Address>>::new(), |mut map, (address, average_missed)| {
+            map.entry(average_missed).or_default().push(address);
+            map
+        });
+    for validators in sorted_validators.values() {
+        let reward = reduced_rewards / (u64::try_from(validators.len()).unwrap() + 1);
+        if reward == 0 {
+            break
```

Comment:
What happens when `reduced_rewards < (u64::try_from(validators.len()).unwrap() + 1)` 
Should we give the remainder to someone?

Reply:
No. we burns the remaining CCC.